src/ultk/effcomm/rate_distortion.py

```diff
@@ -0,0 +1,191 @@
+    speaker = LiteralSpeaker(language)
+    speaker.weights = rows_zero_to_uniform(speaker.normalized_weights())
+    listener = Listener(language)
+    listener.weights = ib_optimal_decoder(speaker.weights, prior, meaning_dists)
+    return {
+        "encoder": speaker,
+        "decoder": listener,
+    }
+
+
+def ib_encoder_to_point(
+    meaning_dists: np.ndarray,
+    prior: np.ndarray,
+    encoder: np.ndarray,
+    decoder: np.ndarray = None,
+) -> tuple[float]:
+    """Return (complexity, accuracy, comm_cost) IB coordinates.
+
+    Args:
+        meaning_dists: array of shape `(|meanings|, |meanings|)` representing the distribution over world states given meanings.
+
+        prior: array of shape `|meanings|` representing the communicative need distribution
+
+        encoder: array of shape `(|meanings|, |words|)` representing P(W | M)
+
+        decoder: array of shape `(|words|, |meanings|)` representing P(M | W).  By default is None, and the Bayesian optimal decoder will be inferred.
+    """
+    if decoder is None:
+        decoder = ib_optimal_decoder(encoder, prior, meaning_dists)
+
+    complexity = information_cond(prior, encoder)
+    accuracy = MI(meaning_dists @ joint(encoder, prior))
+    distortion = MI(joint(meaning_dists, prior)) - accuracy
+
+    return (complexity, accuracy, distortion)
+
+
+def ib_optimal_decoder(
+    encoder: np.ndarray,
+    prior: np.ndarray,
+    meaning_dists: np.ndarray,
+) -> np.ndarray:
+    """Compute the bayesian optimal decoder. See https://github.com/nogazs/ib-color-naming/blob/master/src/ib_naming_model.py#L40
+
+    Args:
+        encoder: array of shape `(|words|, |meanings|)`
+
+        prior: array of shape `(|meanings|,)`
+
+        meaning_dists: array of shape `(|meanings|, |meanings|)`
+
+    Returns:
+        array of shape `(|words|, |meanings|)` representing the 'optimal' deterministic decoder
+    """
+    return bayes(encoder, prior) @ meaning_dists
+
+
+##############################################################################
+# Estimating bounds for a domain
+##############################################################################
+
+
+def get_ib_bound(
+    prior: np.ndarray,
+    meaning_dists: np.ndarray,
+    *args,
+    betas: np.ndarray = np.logspace(-2, 5, 30),
+    **kwargs,
+) -> list[IBResult]:
+    """Estimate the IB theoretical bound for a domain, specified by a prior over meanings, (perceptually uncertain) meaning distributions.
+
+    Args:
+        meaning_dists: array of shape `(|meanings|, |meanings|)` representing the distribution over world states given meanings.
+
+        prior: array of shape `|meanings|` representing the communicative need distribution
+
+    Returns:
+        a list of `rdot.ba.IBResult` namedtuples.
+    """
+    return IBOptimizer(
+        joint(meaning_dists, prior),
+        betas,
+        *args,
+        **kwargs,
+    ).get_results()
```

Comment:
Very cool!  Wondering if we might want to expose the raw optimizer separate from the results at some place, but that's easy to do in the future if we need to

Reply:
Interested to hear more! In principle I can imagine one might want the IBOptimizer instance to look at in addition to getting a curve, since it does indeed contain a lot of useful data and methods.